rfcs/20220217-visual_regression_testing.md

```diff
@@ -0,0 +1,240 @@
+
+We can still _try_ to find a workaround for this, but if it ends up being too much trouble I'd say just leave it out of the final roadmap for this feature and we can use the built-in Playwright tool. Playwright will output the before and after images for failed snapshot comparisons into a `test-results` folder at the top level of the project, so I figure the worst case scenario is that we could write a little script to walk through that directory and run `pixelmatch` on them to get the diff image.
+
+## Repository size considerations
+
+Storing many images in a Git repository can be problematic. Between using more of our contributor's bandwidth in uploading and downloading images to just generally butting up against the limit of GitHub's repository size limit.
+
+The repository size limit I don't think we'll ever need to worry about. GitHub has a [repository size limit](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-large-files-on-github#repository-size-limits) of 100GB. Playwright screenshots are small (in the kilobytes) so it would take a very long time and tons and tons of screenshots to start getting close to the repo size limit.
+
+However, if we're not careful we could significantly increase the size of our repository which would adversely effect contributors by eating their bandwidth when pushing or pull contributions. There are some strategies we could use if this becomes a problem. For example we could use git's `clean` and `smudge` features to losslessly compress and decompress the snapshot images automatically, though I'm not sure how well that would work[^2]. Before we invested too much in that, however, we'd want to ensure that more typical, though lossy, [git repository size reduction](https://stackoverflow.com/questions/2116778/reduce-git-repository-size#2116892) isn't a better option.
+
+Ultimately I think the best option would be to regularly (once a day) run a scheduled maintenance task to remove any historical snapshots older than a month that are not currently used by the main branch. That would mean we would always have a _reasonable_ historical view of snapshots that change frequently and I frankly doubt snapshot history extending more than a month back would be very useful anyway.
+
+## Cross platform rendering
+
+Browsers will sometimes render minute differences between platforms, even with the exact same webpage. This means that contributors running Linux might generate different screenshots from those running macOS and again from those running Windows, even if they're all using the same browser and version.
+
+To avoid this, other visual regression testing libraries like BackstopJS support running tests inside a docker container and using the browser inside the container. We can do the same with Playwright (it's just not as easy).
+
+I was about to figure out "a way" of doing it. Here's what I came up with:
+
+1. Create a `./bin/visual-regression.js` file with the following:
```

Comment:
I think this approach in general is solid! I've used tactics like it in the past to accomplish similar tasks. My only other thought is that we might get tighter/easier integration if we use a JS Docker API library rather than running a subprocess ourselves. I know very little about the space, but maybe [`node-docker-api`](https://www.npmjs.com/package/node-docker-api) could work? It's definitely not a requirement, but it may help with things like log capture or mounts.

Reply:
Oh hey that's a nice idea, I didn't even consider that. Given that it's a single `docker run` command though, I'm reluctant to add another dependency for it :thinking:

Reply:
Yea, I totally hear that :slightly_smiling_face: I mean, we're doing pretty much [the same thing in the API](https://github.com/WordPress/openverse-api/blob/main/ingestion_server/test/integration_test.py#L299), so it's not a necessary change! I only bring it up because I know these libraries typically tend to handle the text streaming aspects a bit better.